docs/design/2022-03-10-write-bypass-tidb.md

```diff
@@ -0,0 +1,117 @@
+- Checkpoint only supports data sources with checkpointing enabled, Flink uses each checkpoint as a transaction separator, and records commit transaction information through the checkpoint, which combined with TiKV's two-stage commit, can achieve the semantics of exactly once when writing TiDB upstream.
+- MiniBatch has no restriction on data sources, no global transaction semantics, and writes to TiDB with a single Task as the dimension. TiDB will be written when the data reaches a specified number of rows or a certain interval since the last write.
+
+### Data de-duplication
+
+If there is a unique index conflict in the data of the same transaction, there will likely be index inconsistency, so the data of the same transaction needs to be de-duplicated.
+
+It is worth noting that data de-duplication is an indeterminate act and there is no guarantee which data will be retained. If there are certain rules for de-duplication, please ensure that the data is not duplicated by the user.
+- MiniBatch mode uses HashMap de-duplication because the number of rows is small and there is no need to communicate with other nodes.
+- Checkpoint uses KeyedProcessFunction, the same key will be routed to the same function instance, then check the checkpointEntry stored in the global state to determine if it is inside the same checkpoint, if so, de-duplication will be performed.
+- Global Using KeyedProcessFunction, the same key will be routed to the same function instance, then check isExisted in the global state store to determine if the same data already exists, and if so, de-duplicate.
+
+## Procedure design
+
+The procedure design of the three modes shows as follows.
+
+It is worth noting that the box actually indicates the add of Transformation, and the operation of the Transformation is not executed immediately, but later executed in parallel in each TaskManager. Transformation includes Function and Operator. So the flowchart shows the order in which each Transformation is added.
+
+### MiniBatch mode
+
+ProcessElement is executed once when it iterates over each element. As you can see, MiniBatch has no global transaction guarantee and performs a 2pc write to the data whenever the number of data rows reaches a certain threshold.
+
+![image alt text](imgs/mini-batch.png)
+
+### Global
+
+Global prewrites the primary key when the stream is generated, then the ProcessElement only prewrites the Secondary Keys and commits the primary key once all the data has been written.
+
+![image alt text](imgs/global.png)
+
+### Checkpoint
+
+Unlike the previous two implementations, TIDBSinkFunction inherits TwoPhaseCommitSinkFunction, which provides exactly-once semantics. calls to the commit function occur at the checkpoint node, as described in [TwoPhaseCommitSinkFunction](https://nightlies.apache.org/flink/flink-docs-master/api/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.html).
```

Comment:
whether we realize the exactly-once semantics. I don't see the rollback logic in the image

Reply:
It's abort method, which is not mentioned in the image